### PR TITLE
tests: improve coverage of jj git push

### PR DESCRIPTION
**Pull request by higgsd**
Improve coverage of `jj git push`.  This is a big file, so there were a lot of little coverage holes to plug.  Let me know if this needs splitting up any further.

Some coverage items remaining that were not covered:
- Branches requiring `--quiet` I skipped because they aren't too exciting
- Signing disabled should never produce a non-empty list of commits to sign (unreachable)
- RejectedBookmarkUpdateReason values are never printed without a hint (unreachable)
- Remote/local rejection without a reason are non-trivial to generate with the git binary
- Rejecting commits due to name/email placeholder is also non-trival, requires low level git commit construction.

# Checklist

If applicable:

- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (`README.md`, `docs/`, `demos/`)
- [ ] I have updated the config schema (`cli/src/config-schema.json`)
- [ ] I have added/updated tests to cover my changes

### CODE REVIEW

File: cli/tests/test_git_push.rs [hunk 1]
```diff
@@ -84,6 +84,61 @@ fn test_git_push_nothing() {
...
+        .success();
+
+    // select remote based on git.push config
+    let output = work_dir.run_jj([
+        "git",
+        "push",
+        "--config",
+        "git.push=other",
+        "-b",
+        "bookmark1",
```
Comment (yuja) [inline cli/tests/test_git_push.rs]: nit: `--config=git.push=other`, etc. to save vertical space

File: cli/tests/test_git_push.rs [hunk 2]
```diff
@@ -1973,6 +1974,18 @@ fn test_git_push_tracked_vs_all() {
...
+    // Bookmark is tracked but no changes to push
+    work_dir
+        .run_jj(["bookmark", "track", "bookmark1@origin"])
+        .success();
+    let output = work_dir.run_jj(["git", "push", "--tracked"]);
+    insta::assert_snapshot!(output, @r"
+    ------- stderr -------
+    Nothing changed.
+    [EOF]
+    ");
```
Comment (yuja) [inline cli/tests/test_git_push.rs]: nit: I don't think noop `--tracked` is particularly special to test, but I don't feel strongly.

File: cli/tests/test_git_push.rs [hunk 3]
```diff
@@ -1978,6 +1978,81 @@ fn test_git_push_deleted_untracked() {
...
+fn test_git_push_deleted_remote_conflict() {
+    let test_env = TestEnvironment::default();
+    set_up(&test_env);
+    let work_dir = test_env.work_dir("local");
+
+    // track the bookmark, de-colocate so git refs can't "fix" built conflict
+    work_dir
+        .run_jj(["bookmark", "track", "bookmark1@origin"])
+        .success();
+    work_dir.run_jj(["git", "colocation", "disable"]).success();
```
Comment (yuja) [inline cli/tests/test_git_push.rs]: Can you set up non-colocated from the start?

File: cli/tests/test_git_push.rs [hunk 4]
```diff
@@ -1010,6 +1010,36 @@ fn test_git_push_changes() {
     [EOF]
     ");
 
+    // Template producing duplicate bookmarks is allowed, even if the sequence
+    // goes backwards.  Last bookmark "wins".
+    let output = work_dir.run_jj(["log", "-T", "change_id.short() ++ ' ' ++ commit_id.short()"]);
```
Comment (yuja) [inline cli/tests/test_git_push.rs]: This is probably a bug. Can you leave a TODO or FIXME comment?

File: cli/tests/test_git_push.rs [hunk 5]
```diff
@@ -1053,6 +1053,22 @@ fn test_git_push_changes() {
...
     ");
+
+    // Generate a bookmark name with non-utf8 content
+    work_dir.write_file("file", vec![b'\xff']);
+    let output = work_dir.run_jj([
+        "git",
+        "push",
+        "--config=templates.git_push_bookmark=self.diff().git()",
+        "--change=@",
+    ]);
```
Comment (yuja) [inline cli/tests/test_git_push.rs:1056]: nit: I don't think this edge case must be covered by CLI tests, but I'm not against adding one.

File: cli/tests/test_git_push.rs [hunk 6]
```diff
@@ -2330,6 +2331,82 @@ fn test_git_push_to_remote_with_slashes() {
...
+                to_toml_value(UserSettings::USER_NAME_PLACEHOLDER)
+            ),
+            "--config",
+            &format!(
+                "user.email={}",
+                to_toml_value(UserSettings::USER_EMAIL_PLACEHOLDER)
+            ),
+            "--update-author",
+        ])
+        .success();
```
Comment (yuja) [inline cli/tests/test_git_push.rs:2358]: Since this compatibility hack is quite old, maybe it's okay to remove the hack instead of adding tests?

File: cli/tests/test_git_push.rs [hunk 7]
```diff
@@ -2536,6 +2536,63 @@ fn test_git_push_sign_on_push() {
...
     [EOF]
     ");
+
+    // Do not sign commits from other authors or re-sign unnecessarily
+    work_dir
+        .run_jj([
+            "metaedit",
+            "--config",
+            "user.email=someone.else@example.com",
+            "--update-author",
```
Comment (yuja) [inline cli/tests/test_git_push.rs]: nit: use `metaedit --author`?

File: cli/tests/test_git_push.rs (hunk 2 above)
Comment (higgsd) [reply cli/tests/test_git_push.rs]: Not very exciting, but was not hit by the previous `test_git_clone_command.rs` code.

File: cli/tests/test_git_push.rs (hunk 4 above)
Comment (higgsd) [reply cli/tests/test_git_push.rs]: Happy to leave a comment, but what part do you consider a bug?  That duplicate bookmarks are allowed, that the sequence could go backwards?

Comment (martinvonz) [reply cli/tests/test_git_push.rs]: I consider it a bug that duplicate bookmarks are allowed

File: cli/tests/test_git_push.rs (hunk 5 above)
Comment (higgsd) [reply cli/tests/test_git_push.rs:1056]: This case is explicitly handled in `jj git push`, so I figured I'd touch it.

File: cli/tests/test_git_push.rs (hunk 6 above)
Comment (higgsd) [reply cli/tests/test_git_push.rs:2358]: Sure, I can drop this test and make a separate PR for removal.

To clarify, are you proposing removal of the placeholder checks in `jj git push`, or the existence of the placeholders entirely?

Comment (martinvonz) [reply cli/tests/test_git_push.rs:2358]: I would think it's the former

File: cli/tests/test_git_push.rs (hunk 2 above)
Comment (yuja) [reply cli/tests/test_git_push.rs]: Is this different from noop `jj git push --all` for example?

File: cli/tests/test_git_push.rs (hunk 6 above)
Comment (yuja) [reply cli/tests/test_git_push.rs:2358]: > To clarify, are you proposing removal of the placeholder checks in jj git push, or the existence of the placeholders entirely?

I'm not sure what "the existence of the placeholders entirely" means, but I would remove the `PLACEHOLDER` constants and their users. It's unlikely that visible commits still have these placeholder names. In any case, this change should be a separate PR.

File: cli/tests/test_git_push.rs (hunk 2 above)
Comment (higgsd) [reply cli/tests/test_git_push.rs]: There is a lot of functional similarity, but it is a distinct block of code.  Depends what you mean my "different".

Comment (yuja) [reply cli/tests/test_git_push.rs]: I assumed they're all handled by `bookmark_updates.is_empty()`, but maybe I'm wrong.

Anyway, since we have `test_git_push_nothing()`, it's probably better to add this and the other noop tests there.

Comment (higgsd) [reply cli/tests/test_git_push.rs]: I see the confusion.  You are correct - the `bookmark_updates.is_empty()` code path is common.

But the coverage paths I was trying to hit were the various instances of `classify_bookmark_update` returning `Ok(None)`.  With the typical test setup, that is easiest with a noop push, though that doesn't have to be the case.

I can update the comments and commit messages to give more clarity, like `tests: exercise git push --branch with a bookmark needing no update`.

I would prefer to keep these tests collated with the push type, to indicate they are specific to the handling of that type.

Comment (yuja) [reply cli/tests/test_git_push.rs]: In practice, I don't think we can do weird things on `classify_bookmark_update()` return value. What we can reasonably do for `Ok(None)` is to ignore, show warning, or error out.

I don't feel strongly where these tests should be added, but I would add noop `--tracked` to `test_git_push_nothing()` because `--all` and `--tracked` are pretty similar.

EDIT: If you meant various cases that `classify_bookmark_update()` returns `AlreadyMatches`, they should be covered by jj-lib tests.

Comment (higgsd) [reply cli/tests/test_git_push.rs]: Agreed - these are not interesting cases, but at least now the test notes what happens in each case (usually nothing).

I moved the `--tracked` case into `test_git_push_nothing()` as requested.
